### DISCUSSION TITLE
Streaming the response

### DISCUSSION THREAD

**Discussion by kurtextrem** [swup/swup#635] [Feature ideas]
Hi, as far as I understand swup waits until the XHR is done before swapping out the DOM and then does the transition.
However, instead of waiting for the full response, the XHR response could be streamed into a DOM node, e.g. like here: https://github.com/jongacnik/streamjax. A [few](https://jakearchibald.com/2016/streams-ftw/) [years](https://developer.chrome.com/articles/fetch-streaming-requests/) [have passed](https://web.dev/streams/), we no longer have to use the iframe hack. All of those approached lead to an improved performance over XHR + innerHTML (see Jake's article). 

## Swup-next
I hear you say: it makes transitions harder... except that Chromium now supports SPA transitions: https://developer.chrome.com/docs/web-platform/view-transitions/. This is very new tech, but in 2016 Web Streams were also very new tech, fast forward to 2023, they're as old as 3+ years now - so what if we start swup-next, that uses modern streaming with view transitions to achieve a native, reload-free, experience?

### Server
Servers need to implement a check if the request is an XHR request (by e.g. setting a special header in the request, such as `swup-xhr: 1`). 
When that header is set, servers should omit `<head>` and stuff that should not be streamed into the container (e.g. do not send the header etc. again, only the content that should replace the previous content)

### `<head>` stuff
As we omit everything that is not part of the updated body (as we're streaming the response right in), we need a way to update the favicon and title, so we probably should also come up with a response header (`swup-favicon:`, `swup-title`)?

### `<script>` / `<link>`

I'm not sure if streaming those in work with the Web Streams. I'd assume yes -- but with the iframe hack it was slightly more work to get them right (see Jake's original article where he proposed the hack). This needs to be taken into account / tested.

---

What does everything think of this? Is this deviating too much from swup, e.g. would a fork make sense? Or a new project under the Swup umbrella?

**Comment by daun** [comment]
This sounds super-interesting. While I haven't yet worked with XHR streams, the performance improvement sounds like a good thing.

From a cursory reading of the requirements and the caveats in the sjax package, this deviates quite a bit from swup's core assumptions:

- Swup supports replacing multiple containers with differing transitions, which would not work with this
- Swup will only fade in the new page after replacing the content, which we wouldn't know when to do since it might still be streaming
- This will forcibly require some server-side setup, while swup is currently happy with simple static sites
- Body classes and title tags are usually important for SEO and styling

To me, it sounds like this would be a great plugin for lower-level libraries like [htmx](https://github.com/bigskysoftware/htmx) or [alpine](https://github.com/alpinejs/alpine).

I'm happy to be convinced by a PR if you want to give this a try, though!
